In this example, we will see GEPA evolve the whole DSPy program (not just the instruction), including modifying the structure/dataflow of the program. We will use GEPA to tune a simple dspy.ChainOfThought module for MATH questions into a full DSPy program.

In [10]:
import os
from dotenv import load_dotenv

# Load environment variables from .env file
load_dotenv()

# Verify OPENAI_API_KEY is loaded
if not os.getenv("OPENAI_API_KEY"):
    raise ValueError("OPENAI_API_KEY not found in .env file")

In [11]:
import dspy

In [12]:
import random

from dspy.datasets import MATH

dataset = MATH(subset="algebra")

# Shuffle the train and dev sets
random.Random(0).shuffle(dataset.train)
random.Random(0).shuffle(dataset.dev)

print(len(dataset.train), len(dataset.dev), len(dataset.test))

350 350 487


Let's inspect an example from the training set.

In [27]:
example = dataset.train[0]
print("Question:", example.question)
print("Reasoning:", example.reasoning)
print("Answer:", example.answer)

Question: If $f(x) = x^3 - 6x^2 + 3x - 4$, $g(x) = x^3 + 5x^2 + 9x - 2$, then find the constant term of $f(g(x))$.
Reasoning: Since $f(g(x)) = g(x)^3 - 6g(x)^2 + 3g(x) - 4$, it suffices to identify the constant terms of $g(x)^3$, $g(x)^2$, and $g(x)$. When expanding $g(x)^3$, we notice that the only way to obtain the constant term is to multiply the constant term with $g(x)$ by itself multiple (3) times: $(-2) \times (-2) \times (-2) = -8$. Similarly, the constant term of $g(x)^2$ is $(-2) \times (-2) = 4$. The constant term in $g(x)$ is $-2$. Substituting gives $(-8) - 6 \cdot 4 + 3 \cdot (-2) - 4 = -8 - 24 - 6 - 4 = \boxed{-42}$.
Answer: -42


Let's define a simple DSPy program to solve this task.

Unlike dspy.GEPA that can take an instantiated DSPy module as input, here, we want to evolve the full DSPy program. Hence, a candidate here is the source code as string. The seed program does not need to be sophisticated, it just needs to demonstrate what the expected input/output interface is, and possibly the available tools. You can also include any additional information about the environment as a comment.

In [14]:
program_src = """import dspy
program = dspy.ChainOfThought("question -> answer")"""

GEPA interfaces with external frameworks through an adapter. In this case, we integrate GEPA with a DspyAdapter.

In [15]:
from gepa.adapters.dspy_full_program_adapter.full_program_adapter import DspyAdapter

In [16]:
def metric_fn(example, pred, trace=None):
    score = dataset.metric(example, pred)
    if score:
        feedback_text = f"The provided answer '{pred.answer}' is correct."
    else:
        feedback_text = f"The provided answer '{pred.answer}' is incorrect. The correct answer is '{example.answer}'. Here's the step by step solution:\n{example.reasoning}"
    return dspy.Prediction(score=score, feedback=feedback_text)

In [17]:
reflection_lm = dspy.LM(model="openai/gpt-4.1", max_tokens=32000)  # temperature=1
adapter = DspyAdapter(
    task_lm=dspy.LM(model="openai/gpt-4.1-nano", max_tokens=32000),
    metric_fn=metric_fn,
    num_threads=80,
    reflection_lm=lambda x: reflection_lm(x)[0],
)

Let's evaluate the base program

In [21]:
o = adapter.evaluate(dataset.test, {"program": program_src})

2025/09/16 15:26:06 INFO dspy.evaluate.evaluate: Average Metric: 316.0 / 487 (64.9%)


The base program obtains a score of 67.1%

Let's launch the GEPA optimization.

In [ ]:
from gepa import optimize

o = optimize(
    seed_candidate={"program": program_src},
    trainset=dataset.train,
    valset=dataset.dev[:200],
    adapter=adapter,
    reflection_lm=lambda x: reflection_lm(x)[0],
    max_metric_calls=2000,
    display_progress_bar=True,
)

GEPA Optimization:  10%|█         | 200/2000 [00:48<07:19,  4.09rollouts/s]

Iteration 0: Base program full valset score: 0.64
Iteration 1: Selected program 0 score: 0.64
Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [00:10<00:00,  3.37s/it]

2025/09/16 15:27:18 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 1: Proposed new text for program: import dspy
import re
from typing import Literal

class MathQA_Signature(dspy.Signature):
    """
    Solve the given math word problem step by step, showing all reasoning and calculations.
    - Output the reasoning as a clear, step-by-step solution.
    - Output the final answer as a concise number or boxed value only (e.g., '42', '6', '32', or '\\boxed{42}').
    - Do NOT include units, sentences, or explanations in the 'answer' field—just the number or boxed value.
    - If multiple valid answers exist, sum them as requested.
    - Common pitfalls: including sentences in the answer, missing edge cases, or not boxing the answer when required.
    - Example:
        Q: John is 31 years younger than his dad. The sum of their ages is 53 years. How many years old is John's dad?
        Reasoning: Let d = dad's age... (show steps)... d = 42.
        Answer: 42
    - Example:
        Q: If $f(x)=10$, find the sum of all possible values of $x$.


2025/09/16 15:27:49 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
2025/09/16 15:28:29 INFO dspy.evaluate.evaluate: Average Metric: 112.0 / 200 (56.0%)
GEPA Optimization:  20%|██        | 406/2000 [02:10<08:53,  2.99rollouts/s]

Iteration 1: Full valset score for new program: 0.56
Iteration 1: Full train_val score for new program: 0.56
Iteration 1: Individual valset scores for new program: [True, False, True, True, False, False, False, True, False, True, False, False, True, False, False, False, False, False, False, False, False, True, True, True, True, True, True, False, True, True, True, False, False, True, True, True, True, True, False, True, True, False, True, True, False, True, False, True, True, True, False, True, True, False, True, True, True, False, False, False, True, False, False, True, False, False, True, False, True, True, True, False, True, True, True, True, True, False, True, False, False, True, True, False, True, True, False, True, True, True, False, True, True, False, True, True, True, False, False, True, False, False, True, True, True, True, True, False, False, False, False, False, True, True, False, False, False, False, True, True, True, False, False, True, True, True, True, True, False, False

2025/09/16 15:28:34 INFO dspy.evaluate.evaluate: Average Metric: 0.0 / 3 (0.0%)



Iteration 2: Proposed new text for program: import dspy
import re
from typing import Optional

class MathQA_Signature(dspy.Signature):
    """
    Solve the given math word problem step by step, showing all reasoning and calculations.
    - Output the reasoning as a clear, step-by-step solution.
    - Output the final answer in the required format:
        * If the answer is a number, output only the number or boxed value (e.g., '42', '\\boxed{42}').
        * If the answer is an interval or union of intervals, output in interval notation, boxed if possible (e.g., '[0,1)', '\\boxed{[0,1)}').
        * If the answer is an algebraic expression, output the fully simplified expression, boxed if possible (e.g., '6x^2 + 30x + 36', '\\boxed{6x^2 + 30x + 36}').
        * If the answer must be rounded, follow the instructions (e.g., "nearest whole number").
    - Do NOT include units, sentences, or explanations in the 'answer' field—just the required value or expression.
    - Common pitfalls:

2025/09/16 15:29:52 INFO dspy.evaluate.evaluate: Average Metric: 0.0 / 3 (0.0%)
GEPA Optimization:  21%|██        | 412/2000 [03:32<17:26,  1.52rollouts/s]

Iteration 2: New subsample score is not better, skipping
Iteration 3: Selected program 0 score: 0.64
Average Metric: 3.00 / 3 (100.0%): 100%|██████████| 3/3 [00:05<00:00,  1.94s/it]

2025/09/16 15:29:57 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
GEPA Optimization:  21%|██        | 415/2000 [03:38<17:58,  1.47rollouts/s]


Iteration 3: All subsample scores perfect. Skipping.
Iteration 3: Reflective mutation did not propose a new candidate
Iteration 4: Selected program 0 score: 0.64
Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [00:14<00:00,  4.97s/it] 

2025/09/16 15:30:12 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 4: Proposed new text for program: import dspy

class MathQA_Signature(dspy.Signature):
    """
    Solve the given math problem step by step, showing all necessary reasoning and calculations.
    - Carefully identify the type of problem (e.g., algebra, geometry, calculus) and select an appropriate solution strategy.
    - For quadratic or polynomial maximization/minimization, use both vertex formula and completing the square as appropriate, and express answers in exact form (e.g., fractions like 3/2, not decimals like 1.5), unless the question explicitly requests a decimal.
    - For systems of equations, show substitution or elimination steps clearly.
    - For radical or fractional answers, simplify fully and rationalize denominators if needed.
    - Always check for extraneous solutions and verify the answer in the original equation if possible.
    - Common pitfalls: giving decimal approximations when exact values are required, skipping algebraic simplification, or not v

2025/09/16 15:30:37 INFO dspy.evaluate.evaluate: Average Metric: 0.0 / 3 (0.0%)
GEPA Optimization:  21%|██        | 421/2000 [04:18<25:09,  1.05rollouts/s]

Iteration 4: New subsample score is not better, skipping
Iteration 5: Selected program 1 score: 0.56
Average Metric: 3.00 / 3 (100.0%): 100%|██████████| 3/3 [00:36<00:00, 12.10s/it]

2025/09/16 15:31:14 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
GEPA Optimization:  21%|██        | 424/2000 [04:54<34:28,  1.31s/rollouts]


Iteration 5: All subsample scores perfect. Skipping.
Iteration 5: Reflective mutation did not propose a new candidate
Iteration 6: Selected program 0 score: 0.64
Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [00:08<00:00,  2.76s/it] 

2025/09/16 15:31:22 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 6: Proposed new text for program: import dspy
from typing import Optional

class MathQAWithGraphAndDecimal(dspy.Signature):
    """
    Solve the given math problem step by step, showing all intermediate calculations and reasoning.
    - If the problem involves a graph, carefully extract the equation of the line by:
        * Identifying the y-intercept by examining where the line crosses the y-axis.
        * Determining the slope by counting grid units: for every horizontal run, count the corresponding vertical rise.
        * Write the equation in slope-intercept form (y = mx + b).
        * Substitute the given x-value to solve for the unknown.
        * Double-check for common pitfalls: miscounting units, misreading intercepts, or using approximate values instead of exact fractions.
    - If the problem involves repeating decimals, convert to a fraction by:
        * Letting x equal the decimal.
        * Multiplying by powers of 10 to align repeating parts.
        * S

2025/09/16 15:31:55 WARNING dspy.adapters.json_adapter: Failed to use structured output format, falling back to JSON mode.
2025/09/16 15:31:58 WARNING dspy.adapters.json_adapter: Failed to use structured output format, falling back to JSON mode.
2025/09/16 15:31:58 ERROR dspy.utils.parallelizer: Error for Example({'question': 'If $f(x)=\\dfrac{a}{x+2}$, solve for the value of $a$ so that $f(0)=f^{-1}(3a)$.', 'reasoning': "The definition of $f$ lets us evaluate $f(0)$: \\[f(0)=\\frac{a}{0+2}=\\frac a{2}.\\]Therefore we want to find all possible $a$ for which \\[\\frac a2=f^{-1}(3a).\\]This is equivalent to  \\[f\\left(\\frac a2\\right)=3a.\\]When we substitute $x=\\frac a2$ into the definition of $f$ we get  \\[f\\left(\\frac a2\\right)=\\frac{a}{\\frac a2+2}=\\frac{2a}{a+4},\\]so we are looking for all solutions $a$ to the equation \\[\\frac{2a}{a+4}=3a.\\]Multiplying both sides by $a + 4$, we get $2a = 3a(a + 4) = 3a^2 + 12a$, so \\[3a^2 + 10a = 0.\\]Then $a(3a + 10) = 0$, so $a = 0$ 

Iteration 6: New subsample score is not better, skipping
Iteration 7: Selected program 0 score: 0.64
Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [00:09<00:00,  3.12s/it]

2025/09/16 15:32:12 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 7: Proposed new text for program: import dspy
import re

class MathQA(dspy.Signature):
    """
    Solve the given math or word problem step by step, showing all intermediate calculations and reasoning.
    - Carefully check each calculation for arithmetic or algebraic errors.
    - For numeric answers, ensure the final answer is formatted as specified in the question (e.g., no commas in numbers, correct rounding, correct units, and no extra text).
    - For coordinate or tuple answers, use the format (x, y) with no extra spaces or text.
    - If the answer is a decimal, round as instructed.
    - Common pitfalls: arithmetic mistakes, copying errors, incorrect formatting (e.g., '978,121' instead of '978121'), or including extra explanation in the answer field.
    - Successful strategy: Show all steps, double-check calculations, and clearly extract the final answer in the required format.
    """
    question: str = dspy.InputField(desc="A math or word problem to solve")
   

2025/09/16 15:32:42 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
2025/09/16 15:33:42 INFO dspy.evaluate.evaluate: Average Metric: 173.0 / 200 (86.5%)
GEPA Optimization:  32%|███▏      | 636/2000 [07:22<17:11,  1.32rollouts/s]

Iteration 7: New program is on the linear pareto front
Iteration 7: Full valset score for new program: 0.865
Iteration 7: Full train_val score for new program: 0.865
Iteration 7: Individual valset scores for new program: [True, True, True, True, False, False, False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, False, True, True, True, False, True, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, False, True, True, True, True, False, False, True, True, False, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, False, True, True, False, True, True, True, True, True, True, False, False, False, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True

2025/09/16 15:33:51 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 8: Proposed new text for program: import dspy
import re

class MathQA(dspy.Signature):
    """
    Solve the given math problem step by step, showing all necessary reasoning and calculations.
    - First, provide a clear, detailed reasoning chain, including algebraic manipulations, substitutions, and checks for extraneous solutions.
    - Then, extract the final answer in the required format (e.g., as a number, fraction, or radical in simplest form).
    - Do NOT enclose the answer in LaTeX boxes (e.g., do not use \boxed{} or similar).
    - For radical or fractional answers, use plain text (e.g., 3*sqrt(3)/4).
    - If multiple solutions, separate them with commas.
    - Common pitfalls: Do not skip domain checks, always verify extraneous solutions, and ensure the answer matches the question's requested form.
    - Edge cases: If the answer is not real or no solution exists, state "No real solution".
    """
    question: str = dspy.InputField(desc="The math problem to solv

2025/09/16 15:34:20 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)
GEPA Optimization:  32%|███▏      | 642/2000 [08:01<21:18,  1.06rollouts/s]

Iteration 8: New subsample score is not better, skipping
Iteration 9: Selected program 0 score: 0.64
Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [00:07<00:00,  2.43s/it] 

2025/09/16 15:34:28 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 9: Proposed new text for program: import dspy
import re
from fractions import Fraction

class MathWordProblemSignature(dspy.Signature):
    """
    Solve math word problems step by step, showing all intermediate reasoning and calculations.
    - First, restate the problem and identify all given quantities and what is being asked.
    - Use relevant algebraic identities, formulas, or proportional reasoning as appropriate.
    - Show all algebraic manipulations, substitutions, and simplifications explicitly.
    - For problems involving fractions, always reduce to lowest terms.
    - For answers, express as a common fraction or integer as required, and box the final answer (e.g., \\boxed{...}).
    - Common pitfalls: arithmetic mistakes, algebraic sign errors, not reducing fractions, misapplying formulas.
    - Edge cases: division by zero, extraneous solutions, ambiguous units.
    - Successful strategies: write out all steps, check each calculation, and verify the answer mat

2025/09/16 15:35:08 INFO dspy.evaluate.evaluate: Average Metric: 0.0 / 3 (0.0%)
GEPA Optimization:  32%|███▏      | 648/2000 [08:49<28:12,  1.25s/rollouts]

Iteration 9: New subsample score is not better, skipping
Iteration 10: Selected program 0 score: 0.64
Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [00:14<00:00,  4.67s/it] 

2025/09/16 15:35:22 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 10: Proposed new text for program: import dspy
from typing import Literal

class MathQA(dspy.Signature):
    """
    Solve the given math problem step by step, providing clear reasoning and a final answer.
    - For range questions, express the answer in standard interval notation, using only parentheses and the union symbol (e.g., (-infty,0)\cup (0,infty)), with no LaTeX, no extra parentheses, and no \boxed.
    - For numeric answers, output only the integer or decimal, with no extra formatting.
    - Do not include LaTeX, dollar signs, or boxed answers in the final answer field.
    - Common pitfalls: Do not copy the answer from the question; do not add extra symbols or formatting; ensure the answer is in the required notation.
    - Edge cases: If the answer is a union of intervals, use '\cup' with no spaces; if the answer is a single interval, write it plainly.
    - Successful strategy: Show all reasoning in the 'reasoning' field, and output only the final answer in the

2025/09/16 15:35:55 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)
GEPA Optimization:  33%|███▎      | 654/2000 [09:36<36:50,  1.64s/rollouts]

Iteration 10: New subsample score is not better, skipping
Iteration 11: Selected program 2 score: 0.865
Average Metric: 3.00 / 3 (100.0%): 100%|██████████| 3/3 [00:05<00:00,  1.75s/it]

2025/09/16 15:36:01 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
GEPA Optimization:  33%|███▎      | 657/2000 [09:41<36:53,  1.65s/rollouts]


Iteration 11: All subsample scores perfect. Skipping.
Iteration 11: Reflective mutation did not propose a new candidate
Iteration 12: Selected program 1 score: 0.56
Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [00:06<00:00,  2.25s/it] 

2025/09/16 15:36:07 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 12: Proposed new text for program: import dspy
import re
from typing import Literal

class MathQA_Signature(dspy.Signature):
    """
    Solve the given math word problem step by step, showing all reasoning and calculations.
    - Output the reasoning as a clear, step-by-step solution.
    - Do NOT include the final answer in the reasoning as a sentence; instead, end with a clear boxed or standalone value if possible.
    - Output the final answer as a concise number, common fraction, or boxed value only (e.g., '42', '6', '\\boxed{42}', '-\\frac{35}{9}').
    - Do NOT include units, sentences, or explanations in the 'answer' field—just the number, fraction, or boxed value.
    - If multiple valid answers exist, sum them as requested.
    - Common pitfalls: including sentences or explanations in the answer, missing edge cases, not boxing the answer when required, or extracting the wrong number.
    - For fractions, use LaTeX format (e.g., '-\\frac{35}{9}').
    - Example:
   

2025/09/16 15:36:46 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
2025/09/16 15:37:51 INFO dspy.evaluate.evaluate: Average Metric: 169.0 / 200 (84.5%)
GEPA Optimization:  43%|████▎     | 863/2000 [11:32<14:22,  1.32rollouts/s]

Iteration 12: Full valset score for new program: 0.845
Iteration 12: Full train_val score for new program: 0.845
Iteration 12: Individual valset scores for new program: [True, True, True, True, False, False, False, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, False, True, True, True, False, True, True, False, True, True, True, True, True, True, True, False, True, True, True, True, True, False, True, True, True, False, True, True, True, True, True, False, True, True, False, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, False, False, False, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True, True, False, True, False, True, True, False, Fals

2025/09/16 15:38:01 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 13: Proposed new text for program: import dspy
import re

class MathQA(dspy.Signature):
    """
    Solve the given math or word problem step by step, showing all intermediate calculations and reasoning.
    - Carefully check each calculation for arithmetic or algebraic errors.
    - For numeric answers, ensure the final answer is formatted as specified in the question (e.g., no commas in numbers, correct rounding, correct units, and no extra text).
    - For coordinate or tuple answers, use the format (x, y) with no extra spaces or text.
    - For interval notation, use square brackets for inclusive bounds, round brackets for exclusive, and no extra text (e.g., [-2, 7]).
    - If the answer is a decimal, round as instructed.
    - Common pitfalls: arithmetic mistakes, copying errors, incorrect formatting (e.g., '978,121' instead of '978121'), including extra explanation in the answer field, or incorrect interval/set notation (e.g., missing brackets, extra text like 'x in').

2025/09/16 15:38:42 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)
GEPA Optimization:  43%|████▎     | 869/2000 [12:23<19:01,  1.01s/rollouts]

Iteration 13: New subsample score is not better, skipping
Iteration 14: Selected program 2 score: 0.865
Average Metric: 3.00 / 3 (100.0%): 100%|██████████| 3/3 [00:20<00:00,  6.70s/it]

2025/09/16 15:39:02 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
GEPA Optimization:  44%|████▎     | 872/2000 [12:43<21:23,  1.14s/rollouts]


Iteration 14: All subsample scores perfect. Skipping.
Iteration 14: Reflective mutation did not propose a new candidate
Iteration 15: Selected program 3 score: 0.845
Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [00:12<00:00,  4.25s/it] 

2025/09/16 15:39:15 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 15: Proposed new text for program: import dspy
import re
from typing import Literal

class MathQA_Signature(dspy.Signature):
    """
    Solve the given math word problem step by step, showing all reasoning and calculations.
    - Output the reasoning as a clear, step-by-step solution.
    - Do NOT include the final answer in the reasoning as a sentence; instead, end with a clear boxed or standalone value if possible.
    - Output the final answer as a concise number, common fraction, or boxed value only (e.g., '42', '6', '\\boxed{42}', '-\\frac{35}{9}', '\\sqrt{10}').
    - Do NOT include units, sentences, or explanations in the 'answer' field—just the number, fraction, or boxed value.
    - If multiple valid answers exist, sum them as requested.
    - Common pitfalls: including sentences or explanations in the answer, missing edge cases, not boxing the answer when required, or extracting the wrong number.
    - For fractions, use LaTeX format (e.g., '-\\frac{35}{9}').
    

2025/09/16 15:40:10 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)
GEPA Optimization:  44%|████▍     | 878/2000 [13:51<32:49,  1.76s/rollouts]

Iteration 15: New subsample score is not better, skipping
Iteration 16: Selected program 0 score: 0.64
Average Metric: 1.00 / 3 (33.3%): 100%|██████████| 3/3 [00:06<00:00,  2.18s/it] 

2025/09/16 15:40:17 INFO dspy.evaluate.evaluate: Average Metric: 1.0 / 3 (33.3%)



Iteration 16: Proposed new text for program: import dspy
from typing import Optional
import re

class MathQA(dspy.Signature):
    """
    Solve the given math problem step by step, showing clear reasoning.
    - First, restate the problem and identify key variables and relationships.
    - Show all intermediate steps and calculations.
    - At the end, clearly state the final answer in plain text or as a reduced fraction/number, not in LaTeX or math mode.
    - Common pitfalls: Do not return the answer in LaTeX (e.g., '\(27\)' or '\frac{7}{25}'). Instead, use plain text (e.g., '27' or '7/25').
    - If the answer is a fraction, reduce it to lowest terms and write as 'a/b'.
    - If the answer is an integer, write as a plain number.
    - Do not include extra formatting, boxes, or math delimiters in the answer field.
    - Only the answer field should contain the final answer; reasoning should be detailed and stepwise.
    """
    question: str = dspy.InputField(desc="The math problem 

2025/09/16 15:40:46 INFO dspy.evaluate.evaluate: Average Metric: 0.0 / 3 (0.0%)
GEPA Optimization:  44%|████▍     | 884/2000 [14:27<38:45,  2.08s/rollouts]

Iteration 16: New subsample score is not better, skipping
Iteration 17: Selected program 2 score: 0.865
Average Metric: 3.00 / 3 (100.0%): 100%|██████████| 3/3 [00:09<00:00,  3.23s/it]

2025/09/16 15:40:56 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
GEPA Optimization:  44%|████▍     | 887/2000 [14:37<39:50,  2.15s/rollouts]


Iteration 17: All subsample scores perfect. Skipping.
Iteration 17: Reflective mutation did not propose a new candidate
Iteration 18: Selected program 0 score: 0.64
Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [00:03<00:00,  1.09s/it] 

2025/09/16 15:40:59 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 18: Proposed new text for program: import dspy
from typing import Optional
from pydantic import BaseModel, Field

class MathQA(dspy.Signature):
    """
    Solve the given math problem step by step. 
    1. Carefully extract all numbers, variables, and operations from the question.
    2. Write a clear, step-by-step reasoning in plain English, showing all intermediate steps and calculations.
    3. Generate a valid, minimal Python code snippet that computes the final answer (use variables as needed, no print statements).
    4. Common pitfalls: 
        - Do not hallucinate numbers or steps.
        - Always use the numbers and operations exactly as stated.
        - For currency, output the answer as a number (e.g., 40), not as a string with a dollar sign, unless the question explicitly asks for a formatted string.
        - For fractions or decimals, round to two decimal places if not exact.
    5. Edge cases: 
        - If the question involves units, include them in the 

2025/09/16 15:41:20 INFO dspy.evaluate.evaluate: Average Metric: 0.0 / 3 (0.0%)
GEPA Optimization:  45%|████▍     | 893/2000 [15:01<44:18,  2.40s/rollouts]

Iteration 18: New subsample score is not better, skipping
Iteration 19: Selected program 3 score: 0.845
Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [00:10<00:00,  3.34s/it] 

2025/09/16 15:41:30 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 19: Proposed new text for program: import dspy
import re
from typing import Literal

class MathQA_Signature(dspy.Signature):
    """
    Solve the given math word problem step by step, showing all reasoning and calculations.
    - Output the reasoning as a clear, step-by-step solution.
    - Do NOT include the final answer in the reasoning as a sentence; instead, end with a clear boxed or standalone value if possible.
    - Output the final answer as a concise number, common fraction, boxed value, point (e.g., '(9,11)'), variable, or set, as appropriate.
    - Do NOT include units, sentences, or explanations in the 'answer' field—just the number, fraction, boxed value, point, variable, or set.
    - If multiple valid answers exist, sum them as requested.
    - Common pitfalls: including sentences or explanations in the answer, missing edge cases, not boxing the answer when required, or extracting the wrong value.
    - For fractions, use LaTeX format (e.g., '-\\frac{35}{9}')

2025/09/16 15:42:43 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
2025/09/16 15:43:38 INFO dspy.evaluate.evaluate: Average Metric: 172.0 / 200 (86.0%)
GEPA Optimization:  55%|█████▍    | 1099/2000 [17:19<13:31,  1.11rollouts/s]

Iteration 19: Full valset score for new program: 0.86
Iteration 19: Full train_val score for new program: 0.86
Iteration 19: Individual valset scores for new program: [True, False, True, True, False, False, True, True, True, True, True, False, True, True, True, False, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, False, True, True, True, True, True, True, True, False, True, True, True, True, True, True, False, False, True, True, False, False, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, False, True, True, False, True, False, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, False, False, False, False, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True,

2025/09/16 15:43:45 INFO dspy.evaluate.evaluate: Average Metric: 1.0 / 3 (33.3%)



Iteration 20: Proposed new text for program: import dspy
import re

class MathQA(dspy.Signature):
    """
    Solve the given math or word problem step by step, showing all intermediate calculations and reasoning.
    - Carefully check each calculation for arithmetic or algebraic errors.
    - For numeric answers, ensure the final answer is formatted as specified in the question (e.g., no commas in numbers, correct rounding, correct units, and no extra text).
    - For fractions, radicals, or expressions, use standard LaTeX math notation (e.g., '\\frac{a}{b}', '\\sqrt{n}', '3\\sqrt{13}').
    - For coordinate or tuple answers, use the format (x, y) with no extra spaces or text.
    - If the answer is a decimal, round as instructed.
    - Common pitfalls: arithmetic mistakes, copying errors, incorrect formatting (e.g., '978,121' instead of '978121'), or including extra explanation in the answer field.
    - Successful strategy: Show all steps, double-check calculations, and clearly ext

2025/09/16 15:44:31 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)
2025/09/16 15:49:28 INFO dspy.evaluate.evaluate: Average Metric: 185.0 / 200 (92.5%)
GEPA Optimization:  65%|██████▌   | 1305/2000 [23:09<15:32,  1.34s/rollouts]

Iteration 20: New program is on the linear pareto front
Iteration 20: Full valset score for new program: 0.925
Iteration 20: Full train_val score for new program: 0.925
Iteration 20: Individual valset scores for new program: [True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, False, False, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, False, True, True, False, True, True, True, True, False, False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, Tr

2025/09/16 15:49:33 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
GEPA Optimization:  65%|██████▌   | 1308/2000 [23:14<15:31,  1.35s/rollouts]


Iteration 21: All subsample scores perfect. Skipping.
Iteration 21: Reflective mutation did not propose a new candidate
Iteration 22: Selected program 4 score: 0.86
Average Metric: 3.00 / 3 (100.0%): 100%|██████████| 3/3 [00:08<00:00,  2.90s/it]

2025/09/16 15:49:42 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
GEPA Optimization:  66%|██████▌   | 1311/2000 [23:23<15:45,  1.37s/rollouts]


Iteration 22: All subsample scores perfect. Skipping.
Iteration 22: Reflective mutation did not propose a new candidate
Iteration 23: Selected program 2 score: 0.865
Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [00:05<00:00,  1.80s/it] 

2025/09/16 15:49:48 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 23: Proposed new text for program: import dspy
import re

class MathQA(dspy.Signature):
    """
    Solve the given math or word problem step by step, showing all intermediate calculations and reasoning.
    - Carefully check each calculation for arithmetic or algebraic errors.
    - For numeric answers, ensure the final answer is formatted as specified in the question (e.g., no commas in numbers, correct rounding, correct units, and no extra text).
    - For interval, set, or fraction answers, use proper LaTeX formatting (e.g., \left(\frac{3}{5},\frac{8}{3}\right]) and reduce all fractions.
    - Do NOT use informal interval notation like (3/5, 8/3]; always use LaTeX with \frac and \left( \right] etc.
    - For coordinate or tuple answers, use the format (x, y) with no extra spaces or text.
    - If the answer is a decimal, round as instructed.
    - Common pitfalls: arithmetic mistakes, copying errors, incorrect formatting (e.g., '978,121' instead of '978121'), using infor

2025/09/16 15:50:29 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
2025/09/16 15:52:11 INFO dspy.evaluate.evaluate: Average Metric: 180.0 / 200 (90.0%)
GEPA Optimization:  76%|███████▌  | 1517/2000 [25:52<07:50,  1.03rollouts/s]

Iteration 23: Full valset score for new program: 0.9
Iteration 23: Full train_val score for new program: 0.9
Iteration 23: Individual valset scores for new program: [True, True, True, True, False, False, False, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, False, False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, False, True, True, False, True, True, True, False, False, False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, Tru

2025/09/16 15:53:01 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
GEPA Optimization:  76%|███████▌  | 1520/2000 [26:42<09:22,  1.17s/rollouts]


Iteration 24: All subsample scores perfect. Skipping.
Iteration 24: Reflective mutation did not propose a new candidate
Iteration 25: Selected program 6 score: 0.9
  0%|          | 0/3 [00:00<?, ?it/s]

2025/09/16 15:53:03 WARNING dspy.clients.lm: LM response was truncated due to exceeding max_tokens=32000. You can inspect the latest LM interactions with `dspy.inspect_history()`. To avoid truncation, consider passing a larger max_tokens when setting up dspy.LM. You may also consider increasing the temperature (currently 0.0)  if the reason for truncation is repetition.


Average Metric: 3.00 / 3 (100.0%): 100%|██████████| 3/3 [00:05<00:00,  1.89s/it]

2025/09/16 15:53:07 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
GEPA Optimization:  76%|███████▌  | 1523/2000 [26:47<09:25,  1.19s/rollouts]


Iteration 25: All subsample scores perfect. Skipping.
Iteration 25: Reflective mutation did not propose a new candidate
Iteration 26: Selected program 0 score: 0.64
Average Metric: 3.00 / 3 (100.0%): 100%|██████████| 3/3 [00:08<00:00,  2.96s/it]

2025/09/16 15:53:15 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
GEPA Optimization:  76%|███████▋  | 1526/2000 [26:56<09:42,  1.23s/rollouts]


Iteration 26: All subsample scores perfect. Skipping.
Iteration 26: Reflective mutation did not propose a new candidate
Iteration 27: Selected program 2 score: 0.865
Average Metric: 3.00 / 3 (100.0%): 100%|██████████| 3/3 [00:10<00:00,  3.52s/it]

2025/09/16 15:53:26 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
GEPA Optimization:  76%|███████▋  | 1529/2000 [27:07<10:15,  1.31s/rollouts]


Iteration 27: All subsample scores perfect. Skipping.
Iteration 27: Reflective mutation did not propose a new candidate
Iteration 28: Selected program 4 score: 0.86
Average Metric: 3.00 / 3 (100.0%): 100%|██████████| 3/3 [00:05<00:00,  1.79s/it]

2025/09/16 15:53:31 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
GEPA Optimization:  77%|███████▋  | 1532/2000 [27:12<10:22,  1.33s/rollouts]


Iteration 28: All subsample scores perfect. Skipping.
Iteration 28: Reflective mutation did not propose a new candidate
Iteration 29: Selected program 6 score: 0.9
Average Metric: 3.00 / 3 (100.0%): 100%|██████████| 3/3 [00:05<00:00,  1.95s/it]

2025/09/16 15:53:37 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
GEPA Optimization:  77%|███████▋  | 1535/2000 [27:18<10:37,  1.37s/rollouts]


Iteration 29: All subsample scores perfect. Skipping.
Iteration 29: Reflective mutation did not propose a new candidate
Iteration 30: Selected program 4 score: 0.86
Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [00:21<00:00,  7.07s/it] 

2025/09/16 15:53:59 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 30: Proposed new text for program: import dspy
import re
from typing import List, Tuple

class MathQA_Signature(dspy.Signature):
    """
    Solve the given math word problem step by step, showing all reasoning and calculations.
    - If the problem involves invertibility of a function modified by a linear term (e.g., f(x)+cx), explicitly:
        - Extract all segment slopes from the graph or description.
        - For invertibility, require that after adding c, all resulting slopes are strictly positive (for c>0) or strictly negative (for c<0).
        - The smallest positive integer b is the minimal integer so that all slopes+b > 0.
        - The largest negative integer a is the maximal integer so that all slopes+a < 0.
        - Compute a^2 + b^2 accordingly.
        - Avoid only checking for nonzero derivatives; require all slopes to have the same sign.
    - Output the reasoning as a clear, step-by-step solution.
    - Do NOT include the final answer in the reasoning 

2025/09/16 15:55:20 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)
GEPA Optimization:  77%|███████▋  | 1541/2000 [29:01<28:37,  3.74s/rollouts]

Iteration 30: New subsample score is not better, skipping
Iteration 31: Selected program 2 score: 0.865
Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [00:06<00:00,  2.13s/it] 

2025/09/16 15:55:27 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)


2025/09/16 15:55:31 WARNING dspy.clients.lm: LM response was truncated due to exceeding max_tokens=32000. You can inspect the latest LM interactions with `dspy.inspect_history()`. To avoid truncation, consider passing a larger max_tokens when setting up dspy.LM. You may also consider increasing the temperature (currently 0.0)  if the reason for truncation is repetition.


Iteration 31: Proposed new text for program: import dspy
import re
from typing import Literal

class MathQA(dspy.Signature):
    """
    Solve the given math or word problem step by step, showing all intermediate calculations and reasoning.
    - Carefully check each calculation for arithmetic or algebraic errors.
    - For numeric answers, ensure the final answer is formatted as specified in the question (e.g., unreduced fraction, decimal, integer, coordinate, etc.).
    - For fractions, use LaTeX format '\\frac{numerator}{denominator}' unless the question requests a decimal.
    - For decimals, round as instructed.
    - For coordinates or tuples, use the format (x, y) with no extra spaces or text.
    - Do not include extra explanation or units in the answer field unless explicitly required.
    - Common pitfalls: arithmetic mistakes, copying errors, incorrect formatting (e.g., decimal instead of fraction, or vice versa), or including extra explanation in the answer field.
    - Suc

2025/09/16 15:56:09 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
2025/09/16 15:57:40 INFO dspy.evaluate.evaluate: Average Metric: 170.0 / 200 (85.0%)
GEPA Optimization:  87%|████████▋ | 1747/2000 [31:21<04:24,  1.05s/rollouts]

Iteration 31: Full valset score for new program: 0.85
Iteration 31: Full train_val score for new program: 0.85
Iteration 31: Individual valset scores for new program: [True, True, True, True, False, False, False, True, True, True, True, True, True, True, False, True, True, True, True, True, False, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, False, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, False, True, True, True, True, False, True, True, True, False, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True, False, True, True, False, True, True, True, False, True, True, True, False, False, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, Tr

2025/09/16 15:57:51 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
GEPA Optimization:  88%|████████▊ | 1750/2000 [31:31<04:31,  1.09s/rollouts]


Iteration 32: All subsample scores perfect. Skipping.
Iteration 32: Reflective mutation did not propose a new candidate
Iteration 33: Selected program 0 score: 0.64
Average Metric: 1.00 / 3 (33.3%): 100%|██████████| 3/3 [00:09<00:00,  3.01s/it] 

2025/09/16 15:58:00 INFO dspy.evaluate.evaluate: Average Metric: 1.0 / 3 (33.3%)



Iteration 33: Proposed new text for program: import dspy
import re

class MathQA(dspy.Signature):
    """
    Solve the given math problem and provide a concise, direct answer in the required format.
    
    Instructions:
    - Read the problem carefully and solve it step by step, showing reasoning.
    - For the 'answer' field, strictly follow the format requested in the question (e.g., use a comma, space, or interval notation as specified).
    - Do NOT include LaTeX delimiters (e.g., \boxed{}, \( \), \[ \], or $ $) or any extra symbols in the 'answer' field.
    - Do NOT include units, explanations, or any text in the 'answer' field unless explicitly requested.
    - If the answer is a list or pair, use the delimiter and order specified in the question (e.g., "a,b" or "a b").
    - For interval notation, use standard mathematical notation without LaTeX (e.g., (-infty, -14)\cup(-3, infty)).
    - Common pitfalls: including LaTeX, extra spaces, or incorrect delimiters; not matching 

2025/09/16 15:58:57 ERROR dspy.utils.parallelizer: Error for Example({'question': 'Find all $x$ that satisfy the inequality $(2x+10)(x+3)<(3x+9)(x+8)$. Express your answer in interval notation.', 'reasoning': 'We have \\begin{align*} (2x+10)(x+3)&<(3x+9)(x+8) \\quad \\Rightarrow\n\\\\ 2(x+5)(x+3)&<3(x+3)(x+8) \\quad \\Rightarrow\n\\\\ 2(x+5)(x+3)-3(x+3)(x+8)&<0 \\quad \\Rightarrow\n\\\\ (2x+10-(3x+24))(x+3)&<0 \\quad \\Rightarrow\n\\\\ (-x-14)(x+3)&<0 \\quad \\Rightarrow\n\\\\ (x+14)(x+3)&>0.\n\\end{align*} This inequality is satisfied if and only if $(x+14)$ and $(x+3)$ are either both positive or both negative.  Both factors are positive for $x>-3$ and both factors are negative for $x<-14$.  When $-14<x<-3$, one factor is positive and the other negative, so their product is negative.   Therefore, the range of $x$ that satisfies the inequality is $ \\boxed{(-\\infty, -14)\\cup(-3,\\infty)} $.', 'answer': '(-\\infty, -14)\\cup(-3,\\infty)'}) (input_keys={'question'}): bad escape \c at 

Iteration 33: New subsample score is not better, skipping
Iteration 34: Selected program 0 score: 0.64
Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [00:08<00:00,  2.99s/it]

2025/09/16 15:59:12 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 34: Proposed new text for program: import dspy
import re
from typing import Optional

class MathQA_Signature(dspy.Signature):
    """
    Solve the given math word problem step by step, showing all necessary algebraic manipulations and logical reasoning in the 'reasoning' field.
    
    - In the 'answer' field, provide ONLY the final numeric answer, with no units, no extra text, and no restatement of the question or answer. 
    - The answer must be a single integer or decimal number, as appropriate.
    - Do NOT include phrases like 'The answer is', 'x =', 'The value of x is', or any explanation in the 'answer' field.
    - Common pitfalls: including extra text, units, or boxed answers; failing to show all steps in reasoning; skipping algebraic manipulations.
    - Edge cases: If multiple numbers are computed, only the final answer to the question should be in 'answer'.
    - Successful strategy: Show all steps in 'reasoning', then clearly state the final answer at the end

2025/09/16 15:59:36 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
2025/09/16 16:00:46 WARNING dspy.utils.parallelizer: SIGINT received. Cancelling.


KeyboardInterrupt: 

2025/09/16 16:03:11 WARNING dspy.clients.lm: LM response was truncated due to exceeding max_tokens=32000. You can inspect the latest LM interactions with `dspy.inspect_history()`. To avoid truncation, consider passing a larger max_tokens when setting up dspy.LM. You may also consider increasing the temperature (currently 0.0)  if the reason for truncation is repetition.
2025/09/16 16:03:11 WARNING dspy.adapters.json_adapter: Failed to use structured output format, falling back to JSON mode.
2025/09/16 16:03:15 ERROR dspy.utils.parallelizer: Error for Example({'question': 'If $\\sqrt{5x}\\cdot\\sqrt{10x}\\cdot\\sqrt{18x}=30$, find $x$.', 'reasoning': 'We begin by combining the radical expressions on the LHS of the equation, and expressing everything in terms of prime factors \\begin{align*} \\sqrt{5x}\\cdot{\\sqrt{10x}}\\cdot{\\sqrt{18x}}& =30\n\\\\\\Rightarrow \\qquad \\sqrt{5\\cdot{x}\\cdot{5}\\cdot{2}\\cdot{x}\\cdot{2}\\cdot{3^2}\\cdot{x}}& =30\n\\\\\\Rightarrow \\qquad \\sqrt{5^2\\c

Let's see the DSPy program found by GEPA

In [23]:
print(o.best_candidate["program"])

AttributeError: 'EvaluationBatch' object has no attribute 'best_candidate'

Evaluating the optimized program

In [ ]:
_ = adapter.evaluate(dataset.test, o.best_candidate)

2025/08/27 20:00:35 INFO dspy.evaluate.evaluate: Average Metric: 454.0 / 487 (93.2%)


We see it going from **67% to 93%** in just a few rounds of optimization!